src/self-update/README.md

```diff
@@ -0,0 +1,89 @@
+
+# Requirements
+
+* Be default self-update shall use Debian Package Manager APT.
+  * NOTE: APT Package Manager provides a very basic rollback mechanism that still can fail (e.g. due to some unexpected device state or power-loss).
+* Device Owner shall be able to use a power-fail safe filesystem and transaction-safe package manager (instead of APT) - if the device support any.
+  * Therefore an extensible plugin shall be provided.
+* In thin-edge package's installation logic no external tools (e.g. grep, sed, ...) shall be used. Instead Rust code/libraries shall be used.
+  * This is to reduce external dependencies and so reduce risc for failures due missing tools or wrong version of tools.
+
+# Expected Pre-Conditions 
+
+The device and it's thereon installed thin-edge has to follow precondition below to allow successful self-update:
+
+1) The current (old) thin-edge on the device must be installed with usual thin-edge installation, or at leat as usual thin-edge installation would do. That is to assure new thin-edge version that is about to be installed has all external and internal files, folders and resources (e.g. config files) available.
+2) Deployment of thin-edge and mosquitto must not be changed manually after initial installation (e.g. moving/renaming installed files, moving/renaming configuration files used by thin-edge, renaming users/groups created during installation, ...).
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Details about the SM Plugin
+
+* A designated Software Management module type - namely "tedge" - will be used manage the self-update.
+
+* The existing SM plugin "apt" will be used and extended to manage that module type "tedge". 
+  * A device owner can implement an own plugin for module type "tedge", e.g. to use any other package manager than APT for self-update. 
+
+* A softlink to be created in `/etc/tedge/sm-plugins/` named `tedge` that points to `apt` plugin, will make the SM agent aware.
+
+* The plugin managing module type "tedge" must take all packages of an update request at once from SM Agent (i.E. Plugin API command "update-list" must be used). That is to get the plugin independent from potential restarts of the SM agent (those restarts are part of updating the SM agent).
```

Comment:
I don't understand.

Reply:
> I don't understand.

Let's rephrase it:

```suggestion
* The plugin managing module type "tedge" must take all packages information of an update request at once from SM Agent (i.E. Plugin API command "update-list" must be used). That is since the SM Agent will restart in the moment when it is updated by the plugin. So the plugin must process the entire update request without any further SM Agent action.```

Does that clarify your question?

Reply:
I got it now.

```suggestion
The agent must request updates on the `tedge` plugin using the `update-list` API of this plugin. So the plugin can process the entire update request without any further SM Agent action. This is required because the agent will be stopped by the plugin.
```

Reply:
This still does not address a state where only some packages succeed. We would have to define a revert operation for partial updates or this will leave the system in the not complete state.
E.g:
* We are on tedge 0.4 (implies tedge-mapper 0.4 and tedge-agent 0.4).
* We request upgrade of following packages: tedge 0.5.1, tedge-agent 0.5.1, tedge-mapper 0.5.2 (even though they are part of the same release the user has to know which versions could be compatible).
* Installation starts but only tedge-mapper succeeds installation to 0..5.1, tedge-agent stays at 0.4

We have a system in rather broken state because there is no way to revert partial upgrade.

I think thin-edge has to be provided as a single package containing multiple binaries and without that we will have to maintain too many special cases.

Reply:
> I got it now.

@didier-wenzek, I just slightly changed your suggestion:

```suggestion
* The agent must request updates on the `tedge` plugin using the `update-list` API of this plugin. So the plugin can process the entire update request without any further SM Agent action. This is required because the agent will be restarted by the plugin.
```

Reply:
> I think thin-edge has to be provided as a single package containing multiple binaries and without that we will have to maintain too many special cases.

Yes, I agree. It really makes sense to re-package thin-edge in a way, that we have a kind of "core package" that can be used for self-update. I will take care to have a separate ticket for that topic.